test/unit/reporters/html/html.test.js

```diff
@@ -70,56 +77,110 @@ describe('HTML Reporter', () => {
+                sandbox.stub(lib, 'currentAbsolutePath').returns('/absolute/report/current/path');
+
+                return emitResult_({currentPath: 'current/path'})
+                    .then(() => {
+                        assert.calledWith(fs.copyAsync, 'current/path', '/absolute/report/current/path');
+                    });
+            });
+
+            it('should save reference image', () => {
+                sandbox.stub(lib, 'referenceAbsolutePath').returns('/absolute/report/reference/path');
+
+                return emitResult_({referencePath: 'reference/path'})
+                    .then(() => {
+                        assert.calledWith(fs.copyAsync, 'reference/path', '/absolute/report/reference/path');
+                    });
+            });
 
-        it('should save current image', () => {
-            sandbox.stub(lib, 'currentAbsolutePath').returns('/absolute/report/current/path');
+            it('should save diff image', () => {
+                const saveDiffTo = sandbox.stub();
+
+                sandbox.stub(lib, 'diffAbsolutePath').returns('/absolute/report/diff/path');
+
+                return emitResult_({saveDiffTo})
+                    .then(() => {
+                        assert.calledWith(saveDiffTo, '/absolute/report/diff/path');
+                    });
+            });
+        });
+    });
 
-            return emitResult_({currentPath: 'current/path'})
-                .then(() => {
-                    assert.calledWith(fs.copyAsync, 'current/path', '/absolute/report/current/path');
-                });
+    describe('without --html-failed-only flag', () => {
+        beforeEach(() => {
+            // calling constructor for its side effect
+            new HtmlReporter(emitter, {failedOnly: false}); // eslint-disable-line no-new
         });
 
-        it('should save reference image', () => {
-            sandbox.stub(lib, 'referenceAbsolutePath').returns('/absolute/report/reference/path');
+        it('should save only reference when screenshots are equal', () => {
+            sandbox.stub(lib, 'referenceAbsolutePath').returns('absolute/reference/path');
 
-            return emitResult_({referencePath: 'reference/path'})
-                .then(() => {
-                    assert.calledWith(fs.copyAsync, 'reference/path', '/absolute/report/reference/path');
-                });
+            emitter.emit(Events.TEST_RESULT, mkStubResult_({
+                referencePath: 'reference/path',
+                equal: true
+            }));
+
+            emitter.emit(Events.END);
+
+            return emitter.emitAndWait(Events.END_RUNNER).then(() => {
+                assert.calledOnce(fs.copyAsync);
+                assert.calledWith(fs.copyAsync, 'reference/path', 'absolute/reference/path');
+            });
         });
 
-        it('should save diff image', () => {
-            const saveDiffTo = sandbox.stub();
+        describe('when screenshots are not equal', () => {
```

Comment:
I left it as it is, since from my perspective, it should be tested with the --html-failed-only and also without it and it should be clear, which one is which one.